manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,627 @@
+        String clientId;
+        String commandTopic;
+
+        public TeltonikaDevice(Topic topic) {
+            this.clientId = topic.tokens.get(1);
+            this.commandTopic = String.format("%s/%s/teltonika/%s/commands",
+                    topicRealm(topic),
+                    this.clientId,
+                    topic.tokens.get(3));
+        }
+    }
+
+    private static final String TELTONIKA_DEVICE_RECEIVE_TOPIC = "data";
+    private static final String TELTONIKA_DEVICE_SEND_TOPIC = "commands";
+    private static final String TELTONIKA_DEVICE_TOKEN = "teltonika";
+    private static final String TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME = "sendToDevice";
+    private static final String TELTONIKA_DEVICE_RECEIVE_COMMAND_ATTRIBUTE_NAME = "response";
+
+    private static final Logger LOG = SyslogCategory.getLogger(API, TeltonikaMQTTHandler.class);
+
+    protected AssetStorageService assetStorageService;
+    protected AssetDatapointService AssetDatapointService;
+    protected TimerService timerService;
+    protected Path DeviceParameterPath;
+
+    protected final ConcurrentMap<String, TeltonikaDevice> connectionSubscriberInfoMap = new ConcurrentHashMap<>();
+
+
+    /**
+     * Indicates if this handler will handle the specified topic; independent of whether it is a publish or subscribe.
+     * Should generally check the third token (index 2) onwards unless {@link #handlesTopic} has been overridden.
+     *
+     */
+    @Override
+    protected boolean topicMatches(Topic topic) {
+        return TELTONIKA_DEVICE_TOKEN.equalsIgnoreCase(topicTokenIndexToString(topic, 2));
+    }
+
+    @Override
+    public void start(Container container) throws Exception {
+        super.start(container);
+        getLogger().info("Starting Teltonika MQTT Handler");
+        ManagerIdentityService identityService = container.getService(ManagerIdentityService.class);
+        assetStorageService = container.getService(AssetStorageService.class);
+        AssetDatapointService = container.getService(AssetDatapointService.class);
+        timerService = container.getService(TimerService.class);
+        DeviceParameterPath = Paths.get("deployment/manager/fleet/FMC003.json");
+        if (!identityService.isKeycloakEnabled()) {
+            getLogger().warning("MQTT connections are supported when not using Keycloak identity provider, only for the Teltonika Telematics devices, until auto-provisioning is fully implemented.");
+            isKeycloak = false;
+        } else {
+            isKeycloak = true;
+            identityProvider = (ManagerKeycloakIdentityProvider) identityService.getIdentityProvider();
+        }
+        //TODO: Figure out a way to update the AssetFilter when a new asset is a candidate to receive a response message
+        // Does it even automatically refresh the Asset list?
+
+        clientEventService.addInternalSubscription(
+                AttributeEvent.class,
+                buildAssetFilter(),
```

Comment:
`AssetFilter` is a bit limited in its' capability I would suggest using `null` and then filtering the incoming `AttributeEvents` for asset type car etc.

Reply:
Could I still use the buildAssetFilter() method's returned AssetFilter using apply() to check if the event applies? This way we retain the same filter functionality but dynamically refresh the applicable asset list on every message.

Reply:
Here is what I am trying to do when entering handleAttributeMessage(): 

`AssetFilter<AttributeEvent> eventFilter = buildAssetFilter();

        if(eventFilter.apply(event) != null) {
            getLogger().finer("eventFilter applied and the event is not for me");
            return;
        }`